tfhe/src/safe_serialization.rs

```diff
@@ -0,0 +1,607 @@
+use std::borrow::Cow;
+use std::fmt::Display;
+
+use crate::conformance::ParameterSetConformant;
+use crate::named::Named;
+use bincode::Options;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use tfhe_versionable::{Unversionize, Versionize};
+
+/// This is the global version of the serialization scheme that is used. This should be updated when
+/// the SerializationHeader is updated.
+const SERIALIZATION_VERSION: &str = "0.5";
+
+/// This is the version of the versioning scheme used to add backward compatibibility on tfhe-rs
+/// types. Similar to SERIALIZATION_VERSION, this number should be increased when the versioning
+/// scheme is upgraded.
+const VERSIONING_VERSION: &str = "0.1";
+
+/// This is the current version of this crate. This is used to be able to reject unversioned data
+/// if they come from a previous version.
+const CRATE_VERSION: &str = concat!(
+    env!("CARGO_PKG_VERSION_MAJOR"),
+    ".",
+    env!("CARGO_PKG_VERSION_MINOR")
+);
+
+/// Tells if this serialized object is versioned or not
+#[derive(Serialize, Deserialize, Copy, Clone, PartialEq, Eq)]
+// This type should not be versioned because it is part of a wrapper of versioned messages.
+#[cfg_attr(tfhe_lints, allow(tfhe_lints::serialize_without_versionize))]
+enum SerializationVersioningMode {
+    /// Serialize with type versioning for backward compatibility
+    Versioned,
+    /// Serialize the type without versioning information
+    Unversioned,
+}
+
+impl Display for SerializationVersioningMode {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Versioned => write!(f, "versioned"),
+            Self::Unversioned => write!(f, "unversioned"),
+        }
+    }
+}
+
+/// `HEADER_LENGTH_LIMIT` is the maximum `SerializationHeader` size which
+/// `DeserializationConfig::deserialize_from` is going to try to read (it returns an error if
+/// it's too big).
+/// It helps prevent an attacker passing a very long header to exhaust memory.
+const HEADER_LENGTH_LIMIT: u64 = 1000;
+
+/// Header with global metadata about the serialized object. This help checking that we are not
+/// deserializing data that we can't handle.
+#[derive(Serialize, Deserialize)]
+// This type should not be versioned because it is part of a wrapper of versioned messages.
+#[cfg_attr(tfhe_lints, allow(tfhe_lints::serialize_without_versionize))]
+struct SerializationHeader {
+    header_version: Cow<'static, str>,
+    versioning_mode: SerializationVersioningMode,
+    versioning_version: Cow<'static, str>,
```

Comment:
Maybe better to have  `versioning_version` in `versioning_mode` as its meaning  depends on its value

We could have 
```
enum SerializationVersioningMode {
    /// Serialize with type versioning for backward compatibility
    Versioned { version: Cow<'static, str>, },
    /// Serialize the type without versioning information
    Unversioned  { crate_version: Cow<'static, str>, },
}
```